### PR TITLE
feat(graphback): upgrade to Graphback 0.15

### PR DESCRIPTION
**Pull request by machi1990**
Opened as draft while waiting for 0.14 release. 

/cc @wtrocki @craicoverflow 

## Progress
- [x] Provide update for commands and basic integration

Once this is merged, there will be followup PRs as described here https://github.com/aerogear/graphback/issues/1736

### CODE REVIEW

**Comment by machi1990** [comment]
Supersedes:
 - https://github.com/Urigo/graphql-cli/commit/410bf491e39ffc3b4c80176289ead1bc0897ce88
 - https://github.com/Urigo/graphql-cli/pull/1302

File: packages/commands/generate/src/index.ts [hunk 1]
```diff
@@ -29,14 +25,8 @@ export default defineCommand<{}, CliFlags>(() => {
         },
       });
     },
-    async handler(args) {
-      if (args.backend) {
-        await generateUsingPlugins(args);
-      } else if (args.db) {
-        await createDBResources({});
```
Comment (wtrocki) [inline packages/commands/generate/src/index.ts:36]: We have moved DB to the templates itself as most of the users want to control when database migration happens and CLI was to restrictive - could not handle ability to connect to remote databases very well. 

New change gives developers more flexibility to specify custom scalars, filter operations etc. that will be hard to represent for CLI

File: packages/commands/generate/src/index.ts [hunk 2]
```diff
@@ -29,14 +25,8 @@ export default defineCommand<{}, CliFlags>(() => {
         },
       });
     },
-    async handler(args) {
-      if (args.backend) {
-        await generateUsingPlugins(args);
```
Comment (wtrocki) [inline packages/commands/generate/src/index.ts:34]: We have now 2 concepts of plugins: Plugins that transform schema at runtime and plugins that affect filesystem - like client side queries

File: templates/fullstack/client/src/graphql/graphback.graphql [hunk 3]
```diff
@@ -0,0 +1,147 @@
+fragment NoteFields on Note {
```
Comment (wtrocki) [inline templates/fullstack/client/src/graphql/graphback.graphql:1]: Users complained about generated queries being mixed with their own ones and their changes being overwritten by generator. After periods of trying things with detecting user changes in the files we decided to go with single graphql file that contains all queries as we always feed that into the GraphQL-Code-Generator. GraphQL-Code-Generator is now core for graphback - it gives hooks and resolver typing that are essential for the developers.

File: templates/fullstack/server/src/graphql.ts [hunk 4]
```diff
@@ -1,26 +1,37 @@
...
+
+  const schema = (await loadFiles(join(__dirname, '../../model/'))).join('\n');
+  const { resolvers, contextCreator, typeDefs} = buildGraphbackAPI(schema, {
+    dataProviderCreator: createKnexDbProvider(db),
+    plugins: [new SchemaCRUDPlugin({
+      outputPath: "./src/schema/schema.graphql"
+    })]
+  });
+
+  migrateDB(dbConfig, typeDefs, {
```
Comment (wtrocki) [inline templates/fullstack/server/src/graphql.ts:25]: Migration is performed on the template using the same configuration. 
We also support MongoDB as database and have written nice UI for both Mongo and Postgress.

File: templates/fullstack/server/src/schema/schema.graphql [hunk 5]
```diff
@@ -1,29 +1,87 @@
 ## NOTE: This schema was generated by Graphback and should not be changed manually
 
+"""Exposes a URL that specifies the behaviour of this scalar."""
```
Comment (wtrocki) [inline templates/fullstack/server/src/schema/schema.graphql:3]: Our approach is schema/code/mapping first at the same time. 
Developers describe their datarequirements using models (that are graphql schema and getting fully featured schema in results.

Workflows are very flexible - do not like graphback resolver layer - generate just schema (using plugins) - Do not like client side queries etc.

File: templates/fullstack/server/src/schema/schema.graphql [hunk 6]
```diff
@@ -1,29 +1,87 @@
...
   noteId: ID
 }
 
+input CreateNoteInput {
+  id: ID
+  title: String!
+  description: String
+}
+
+input IDInput {
```
Comment (wtrocki) [inline templates/fullstack/server/src/schema/schema.graphql:55]: As you see graphback is now fully compatible with the https://graphqlcrud.org matching features of the frameworks like prisma and hasura. GraphqlCrud gives us ability to use many partner libraries that rely on generated schema/client side queries

Comment (wtrocki) [reply templates/fullstack/server/src/schema/schema.graphql:55]: We still did not announced it (as still working some terms with partners and need to improve spec - it is very very rough at the moment)

File: templates/fullstack/server/src/graphql.ts (hunk 4 above)
Comment (ardatan) [reply templates/fullstack/server/src/graphql.ts:25]: So we won't generate code anymore but we will generate the executable schema on runtime instead, right?

Comment (wtrocki) [reply templates/fullstack/server/src/graphql.ts:25]: Do you mean GraphQL resolvers?
There is ultimate flexibility here. Plugins giving ability to generate any code, however by default out templates do not generate resolvers and build them at runtime by default.

Point taken was that if generated resolvers are just stubs and you cannot change then (as subsequent generation will override them again) their value is very minimal.
This is a tradeoff - if we take look into runtime we have there is a lots of logic involved in filtering, pagination, mapping complex queries, connecting to existing databases and tables, ID representation, relationships etc. With graphback, you can generate your own resolvers to connect with custom logic. It brings the best from descoped projects like Prisma 1 and TypeORM and comparing to some knex wrappers like: https://www.npmjs.com/package/graphql-server-crud Graphback offers platform and swiss army knife for CRUD operations that can be integrated with other tools by utilizing https://graphqlcrud.org

File: templates/fullstack/.graphqlrc.yml [hunk 7]
```diff
@@ -34,24 +34,18 @@ extensions:
...
+      findOne: true
       delete: true
       subCreate: true
       subUpdate: true
       subDelete: true
     ## Codegen plugins
     plugins:
         graphback-schema:
-          format: 'graphql'
           outputPath: ./server/src/schema
```
Comment (craicoverflow) [inline templates/fullstack/.graphqlrc.yml:46]: Recommended practice is full path to output schema.

Comment (machi1990) [reply templates/fullstack/.graphqlrc.yml:46]: Good catch!

**Comment by wtrocki** [comment]
Ping @ardatan 
Would you have some time to review this?

**Comment by machi1990** [comment]
Looks like npm is down or something causing CI failures.

**Pull request merged**